pkg/volume/util.go

```diff
@@ -332,3 +332,239 @@ func ChooseZoneForVolume(zones sets.String, pvcName string) string {
+	if pvc.Spec.Selector == nil {
+		return "", fmt.Errorf("missing selector.matchLabels")
+	}
+	if value, ok := pvc.Spec.Selector.MatchLabels[key]; ok {
+		return value, nil
+	}
+	return "", fmt.Errorf("key %q not found in selector.matchLabels", key)
+}
+
+// GetPVCMatchExpression returns:
+// - either ([]setOfValues, nil) for all matching (key, operator) from the matchExpressions Selector part of the PVC
+// - or ([]emptySet, error) in case the operator or the key is missing in the matchExpressions Selector part of the PVC
+// Example:
+// selector:
+//     matchExpressions:
+//       - key: failure-domain.beta.kubernetes.io/zone
+//         operator: In
+//         values:
+//           - us-east-1a
+//           - us-east-2a
+//           - us-east-3a
+//       - key: failure-domain.beta.kubernetes.io/zone
+//             operator: In
+//             values:
+//               - us-east-3a
+//               - us-east-4a
+// Returns (sets.String{"us-east-3a": sets.Empty{}}, nil) because all operators are ANDed.
+func GetPVCMatchExpression(pvc *v1.PersistentVolumeClaim, key string, operator metav1.LabelSelectorOperator) ([]sets.String, error) {
+	if pvc.Spec.Selector == nil {
+		return make([]sets.String, 0), fmt.Errorf("missing selector.matchExpressions")
+	}
+	if len(pvc.Spec.Selector.MatchExpressions) < 1 {
+		return make([]sets.String, 0), fmt.Errorf("key(s), operator(s) and value(s) are missing in selector.matchExpressions")
+	}
+	capacity := 0
+	for _, item := range pvc.Spec.Selector.MatchExpressions {
+		if item.Key == key && item.Operator == operator && len(item.Values) > 0 {
+			capacity++
+		}
+	}
+	if capacity == 0 {
+		return make([]sets.String, 0), fmt.Errorf("operator %q for key %q not found in selector.matchExpressions", key, operator)
+	}
+
+	ret := make([]sets.String, 0, capacity)
+	index := 0
+	for _, item := range pvc.Spec.Selector.MatchExpressions {
+		if item.Key == key && item.Operator == operator && len(item.Values) > 0 {
+			ret = append(ret, make(sets.String))
+			for _, value := range item.Values {
+				ret[index].Insert(value)
+			}
+			index++
+		}
+	}
+	return ret, nil
+}
+
+// Regions2ZonesParams are parameters for the func Regions2Zones
+type Regions2ZonesParams struct {
+	// set of regions to be converted to a set of zones
+	Regions sets.String
+	// all currently available zones
+	AllAvailableZones sets.String
+	// a func that converts a zone to a region
+	Zone2region func(string) (string, error)
+}
+
+// Regions2Zones converts a set of regions to a set of zones
+func Regions2Zones(in Regions2ZonesParams) (sets.String, error) {
+	ret := make(sets.String)
+	for zone := range in.AllAvailableZones {
+		for region := range in.Regions {
+			if regionOfZone, err := in.Zone2region(zone); err != nil {
+				return make(sets.String), fmt.Errorf("failed to convert zone (%v) to a region: %v", zone, err.Error())
+			} else if strings.Compare(regionOfZone, region) == 0 {
```

Comment:
Is this the same as `regionOfZone == region`?

I'm actually very unclear as to what this whole function is actually doing.

Can we just call ZoneToRegion once per zone, and then check if in.Regions `Has` the region?

I could use a comment as to why we're doing this per-region filtering!